### PR TITLE
feat(Pagination): Pagination Component Typescript Conversion

### PR DESCRIPTION
**Pull request by nicolethoen**
Addresses [issue 2115](https://github.com/patternfly/patternfly-react/issues/2115)

Includes new demo and cypress tests

### CODE REVIEW

File: packages/patternfly-4/react-core/src/components/Pagination/PaginationOptionsMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+       </strong>{' '}
+       of<strong>{itemCount}</strong> {itemsTitle}
+     </React.Fragment>
+   ),
+   onPerPageSelect: () => null as any
+ };
+
+  constructor(props: PaginationOptionsMenuProps){
+    super(props);
+    this.state = {
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Pagination/PaginationOptionsMenu.tsx]: You can move `state` out of the constructor and into a class member to also get rid of `PaginationOptionsMenuState`.

File: packages/patternfly-4/react-core/src/components/Pagination/OptionsToggle.tsx [hunk 2]
```diff
@@ -0,0 +1,69 @@
+import * as React from 'react';
+import styles from '@patternfly/react-styles/css/components/OptionsMenu/options-menu';
+import { css, getModifier } from '@patternfly/react-styles';
+
+import { fillTemplate } from '../../helpers';
+import { ToggleTemplateProps } from './ToggleTemplate';
+import { DropdownToggle } from '../Dropdown';
+
+export interface OptionsToggleProps extends React.HTMLProps<HTMLDivElement> {
+  /** The type or title of the items being paginated */
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Pagination/OptionsToggle.tsx]: Thanks for adding these comments!

File: packages/patternfly-4/react-core/src/components/Pagination/OptionsToggle.tsx [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  /** showToggle */
+  showToggle?: boolean,
+  /** Event function that fires when user clicks the Options Menu toggle */
+  onToggle?: (isOpen: boolean) => void,
+  /** Flag indicating if the Options Menu dropdown is open or not */
+  isOpen?: boolean,
+  /** */
+  parentRef?: HTMLElement;
+  /** This will be shown in pagination toggle span. You can use firstIndex, lastIndex, itemCount, itemsTitle props. */
+  toggleTemplate?: ((props: ToggleTemplateProps) => React.ReactElement) | string;
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Pagination/OptionsToggle.tsx]: This is much better than using `React.ReactNode` and then casting to `any`. Great job!

File: packages/patternfly-4/react-core/src/components/Pagination/PaginationOptionsMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+  render() {
+    const { className, widgetId, itemsPerPageTitle, dropDirection, optionsToggle, perPageOptions, toggleTemplate, firstIndex, lastIndex, itemCount, itemsTitle } = this.props;
+    const { isOpen } = this.state;
+    let parentRef = null;
+
+    return (
+      <div
+        className={css(styles.optionsMenu, className)}
+        ref={ref => {
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Pagination/PaginationOptionsMenu.tsx]: Could we make `parentRef` a class property and then just do `ref={parentRef}`?

File: packages/patternfly-4/react-core/src/components/Pagination/PaginationOptionsMenu.tsx (hunk 1 above)
Comment (redallen) [reply packages/patternfly-4/react-core/src/components/Pagination/PaginationOptionsMenu.tsx]: I changed my mind, this is fine.

**Comment by nicolethoen** [comment]
@tlabaj @dlabaj can I get a review from a senior dev on this?

**Comment by nicolethoen** [comment]
@tlabaj @dlabaj nvm I did something wrong when I rebased. I'll ping you again when I have it sorted out

**Comment by nicolethoen** [comment]
@dlabaj @tlabaj okay, sorry for the blast of notifications, but I think it's ready for review now. my rebasing did lose @rebeccaalpert and @redallen's approvals

File: packages/patternfly-4/react-core/src/components/Pagination/Navigation.tsx [hunk 5]
```diff
@@ -1,62 +1,62 @@
...
+  toPreviousPage = 'Go to previous page',
+  currPage = 'Current page',
+  paginationTitle = 'Pagination',
+  onNextClick = () => undefined,
+  onPreviousClick = () => undefined,
+  onFirstClick = () => undefined,
+  onLastClick = () => undefined,
+  onPageInput = () => undefined,
   ...props
 }) => (
```
Comment (redallen) [inline packages/patternfly-4/react-core/src/components/Pagination/Navigation.tsx]: Change to `}: NavigationProps) => (`.

**Comment by dlabrecq** [comment]
I attempted to test this with Cost Management, but the Tippy issue is getting in the way. If you can rebase after the release is fixed, I'll test again.

**Comment by nicolethoen** [comment]
@dlabrecq done

**Comment by redallen** [comment]
Will merge after the latest prerelease gets promoted.

**Pull request merged**